### PR TITLE
Add timestamp support to secondary instance

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR adds timestamp support to the secondary DB instance.

With this, these timestamp related APIs are supported:

ReadOptions.timestamp : read should return the latest data visible to this specified timestamp
Iterator::timestamp() : returns the timestamp associated with the key, value
DB:Get(..., std::string* timestamp) : returns the timestamp associated with the key, value in timestamp

Test plan (on devserver):
```
$COMPILE_WITH_ASAN=1 make -j24 all
$./db_secondary_test --gtest_filter=DBSecondaryTestWithTimestamp*
```

### CODE REVIEW

File: db/db_secondary_test.cc [hunk 1]
```diff
@@ -1254,6 +1261,403 @@ TEST_F(DBSecondaryTest, OpenWithTransactionDB) {
   ASSERT_OK(TryOpenSecondary(options));
 }
 
+class DBSecondaryTestWithTimestamp : public DBSecondaryTestBase {
+ public:
+  DBSecondaryTestWithTimestamp()
```
Comment (riversand963) [inline db/db_secondary_test.cc:1266]: Nit: mark as explicit

File: db/db_secondary_test.cc [hunk 2]
```diff
@@ -116,6 +118,11 @@ void DBSecondaryTest::CheckFileTypeCounts(const std::string& dir,
   ASSERT_EQ(expected_manifest, manifest_cnt);
 }
 
+class DBSecondaryTest : public DBSecondaryTestBase {
+ public:
+  DBSecondaryTest() : DBSecondaryTestBase("db_secondary_test") {}
```
Comment (riversand963) [inline db/db_secondary_test.cc:123]: Nit; mark as explicit

File: db/db_secondary_test.cc [hunk 3]
```diff
@@ -18,18 +19,18 @@
 namespace ROCKSDB_NAMESPACE {
 
 #ifndef ROCKSDB_LITE
-class DBSecondaryTest : public DBTestBase {
+class DBSecondaryTestBase : public DBBasicTestWithTimestampBase {
  public:
-  DBSecondaryTest()
-      : DBTestBase("db_secondary_test", /*env_do_fsync=*/true),
+  DBSecondaryTestBase(const std::string& dbname)
```
Comment (riversand963) [inline db/db_secondary_test.cc:24]: Nit: mark as explicit.

**Review by riversand963 — APPROVED** [review]

File: db/db_with_timestamp_test_util.h [hunk 4]
```diff
@@ -17,7 +17,7 @@ namespace ROCKSDB_NAMESPACE {
 class DBBasicTestWithTimestampBase : public DBTestBase {
  public:
   explicit DBBasicTestWithTimestampBase(const std::string& dbname)
-      : DBTestBase(dbname, /*env_do_fsync=*/true) {}
+      : DBTestBase(dbname, /*env_do_fsync*/ true) {}
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/db_with_timestamp_test_util.h]: Agreed, thanks for the style guide pointer.

**Pull request closed without merging**